### PR TITLE
[JBEAP-23082] Fix warning due to incorrect value of ImageStream apiVe…

### PR DESCRIPTION
**Pull request by soul2zimate**
…rsion.

Issue https://issues.redhat.com/browse/JBEAP-23082

Fix warning due to incorrect value of ImageStream and Template apiVersion since [OCP 4.7 deprecation](https://docs.openshift.com/container-platform/4.7/release_notes/ocp-4-7-release-notes.html#ocp-4-7-apiversion-v1)

### CODE REVIEW

**Review by jmesnil — APPROVED** [review]

**Comment by soul2zimate** [comment]
Hi @jmesnil  @luck3y The JBEAP issue is scheduled for EAP 7.4.5. Can this get merged ?

**Comment by soul2zimate** [comment]
@jmesnil I have updated this to include Template `apiVersion` change as well. 

@mackaypeter please review from QE perspective before merging. Thanks

**Review by luck3y — APPROVED** [review]

**Comment by mackaypeter** [comment]
Should we not also fix this for the resources defined in the templates while we're at it? I know this fixes the reported issue, but there's still lots of services, build configs, app image streams etc. using `"apiVersion": "v1"`

Also are we no longer using dev branches for pretesting the changes before merging them?

**Comment by soul2zimate** [comment]
I think it's fine. I don't see that warning for other resources from templates. Resources have been created without any warning when I create a simple app from template (OpenShift version 4.9.15).

`% oc new-app --template=eap74-basic-s2i -p EAP_IMAGE_NAME=jboss-eap74-openjdk11-openshift:7.4.0 -p EAP_RUNTIME_IMAGE_NAME=jboss-eap74-openjdk11-runtime-openshift:7.4.0 -p SOURCE_REPOSITORY_URL=https://github.com/jboss-developer/jboss-eap-quickstarts -p SOURCE_REPOSITORY_REF=7.4.x -p CONTEXT_DIR=kitchensink `

Outcome:

```
--> Deploying template "chaowan-dev/eap74-basic-s2i" to project chaowan-dev

     JBoss EAP 7.4.0
     ---------
     An example JBoss Enterprise Application Platform application. For more information about using this template, see https://github.com/jboss-container-images/jboss-eap-7-openshift-image/blob/eap74/README.adoc

     A new JBoss EAP based application has been created in your project.

     * With parameters:
        * Application Name=eap-app
        * EAP Image Name=jboss-eap74-openjdk11-openshift:7.4.0
        * EAP Runtime Image Name=jboss-eap74-openjdk11-runtime-openshift:7.4.0
        * Git Repository URL=https://github.com/jboss-developer/jboss-eap-quickstarts
        * Git Reference=7.4.x
        * Context Directory=kitchensink
        * Galleon layers=
        * Enable ExampleDS datasource=false
        * Queues=
        * Topics=
        * AMQ cluster password=cs81tYVQ # generated
        * Github Webhook Secret=aqPXCAAp # generated
        * Generic Webhook Secret=2AU2CTPj # generated
        * ImageStream Namespace=openshift
        * JGroups Cluster Password=1BmVKrq8 # generated
        * Deploy Exploded Archives=false
        * Maven mirror URL=
        * Maven Additional Arguments=-Dcom.redhat.xpaas.repo.jbossorg
        * ARTIFACT_DIR=
        * MEMORY_LIMIT=1Gi

--> Creating resources ...
    service "eap-app" created
    service "eap-app-ping" created
    route.route.openshift.io "eap-app" created
    imagestream.image.openshift.io "eap-app" created
    imagestream.image.openshift.io "eap-app-build-artifacts" created
    buildconfig.build.openshift.io "eap-app-build-artifacts" created
    buildconfig.build.openshift.io "eap-app" created
    deploymentconfig.apps.openshift.io "eap-app" created
--> Success
    Access your application via route 'eap-app-chaowan-dev.apps.sandbox.x8i5.p1.openshiftapps.com' 
    Build scheduled, use 'oc logs -f buildconfig/eap-app-build-artifacts' to track its progress.
    Build scheduled, use 'oc logs -f buildconfig/eap-app' to track its progress.
    Run 'oc status' to view your app.
```

**Comment by mackaypeter** [comment]
@luck3y @jmesnil @soul2zimate I'm still not too happy about merging changes directly into eap74. Was there a decision to drop dev branches I'm not aware of or is this simply the first 74 template change, so we don't have a dev branch yet and should create one?

**Comment by jmesnil** [comment]
@mackaypeter Since EAP 7.4, we've agreed with QE that once the corresponding release was out, the templates would be patched with PR by having approved reviews from both QE and engineering.
We have done so for EAP 7.4, XP 3 and 4.

If QE wants to revisit that process, I'm fine discussing it.

**Comment by mchoma** [comment]
> @mackaypeter Since EAP 7.4, we've agreed with QE that once the corresponding release was out, the templates would be patched with PR by having approved reviews from both QE and engineering. We have done so for EAP 7.4, XP 3 and 4.
> 
> If QE wants to revisit that process, I'm fine discussing it.

I remember we on first initial streams (e.g. some xp) maybe agreed we do not need to create `-dev` branch and optimize it to prepare directly on branch which went "live" after release published. But that still meant we tested before release. And maybe we did something similar with some hotfix, just after release.

But here on CP I would say situation is little bit different and we would benefit here from some standardisation to reduce confusion, as branch is already live. It can happen there will be several fixes in CP. So once CP QE will come to test all changes in scope of CP testing, it can be tested as one bulk. Testing each fix individually is not what we want.

I understand in this particular case if this is only one fix in current CP, it can be technically optimized to avoid `-dev` branch as intermmediate branch. But part of handover email to QE on PM list should be exact template repo and branch. In this case it would be personal repo https://github.com/soul2zimate/jboss-eap-openshift-templates.git and branch there JBEAP-23082-eap74, which could sound little bit ackward for readers I would say. Some `-dev` testing branch would sound more appropriate.

**Comment by jmesnil** [comment]
@mchoma That sounds sensible. As we are moving away from OpenShift Templates, I'd expect that their payload will never be high for remaining CPs anyway.

**Comment by soul2zimate** [comment]
Hi all, What would be the right time to sync `eap74-dev` branch to `eap74` branch?

If a new `eap74-dev` branch would be needed for testing purpose, could someone please create it ?  so I can adjust this to target that new `dev` branch.

**Comment by yselkowitz** [comment]
The imagestreams look fine, but templates typically contain multiple objects, each of which need to be handled, e.g. Route's v1 => route.openshift.io/v1 etc.  Those APIs (e.g. Service) which are documented as core/v1 stay as "v1".

**Review by jmesnil — APPROVED** [review]
@soul2zimate thanks

The 3 resources that continue to use non-groupified `apiVersion` are now `List`, `Service` and `PersistentVolumeClaim`

**Comment by soul2zimate** [comment]
I see that new https://issues.redhat.com/browse/SO-47 is blocked by the linked JBEAP issue. 

Hi @mackaypeter  @mchoma @jmesnil, any idea how to unblock this from being merged ?

**Comment by luck3y** [comment]
Can we merge this? I don't see whats blocking it on https://issues.redhat.com/browse/SO-47.

**Comment by jmesnil** [comment]
@mchoma can we get a simplified QE validation for it to expedite this issue (we are on the clock to do it before OCP 4.11 is out).

QE would signoff on https://github.com/soul2zimate/jboss-eap-openshift-templates.git JBEAP-23082-eap74 and we would merge the PR after signoff

**Comment by yselkowitz** [comment]
Ping?

**Comment by mackaypeter** [comment]
@yselkowitz we tested this with the EAP 7.4.5 images and it can be merged once we release them on Monday

**Comment by yselkowitz** [comment]
@mackaypeter please also be sure to get #49, as SSO 7.4 is no longer supported.

**Pull request merged**
